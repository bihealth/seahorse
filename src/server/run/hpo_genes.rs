//! Implementation of `/hpo/genes`.

use std::collections::HashMap;

use actix_web::{
    get,
    web::{self, Data, Json, Path},
    Responder,
};
use hpo::{
    annotations::{AnnotationId, Gene, GeneId},
    Ontology,
};

use crate::server::run::WebServerData;

use super::{CustomError, Match, ResultHpoTerm};

/// Parameters for `fetch_hpo_genes`.
///
/// This allows to query for genes.  The first given of the following is
/// interpreted.
///
/// - `gene_id` -- specify gene ID (either NCBI or HGNC gene ID)
/// - `gene_symbol` -- specify the gene symbol
/// - `max_results` -- the maximnum number of records to return
/// - `hpo_terms` -- whether to include `"hpo_terms"` in result
///
/// The following propery defines how matches are performed:
///
/// - `match` -- how to match
#[derive(serde::Serialize, serde::Deserialize, utoipa::ToSchema, Debug, Clone)]
struct Query {
    /// The gene ID to search for.
    pub gene_id: Option<String>,
    /// The gene symbol to search for.
    pub gene_symbol: Option<String>,
    /// The match mode.
    #[serde(alias = "match")]
    pub match_: Option<Match>,
    /// Maximal number of results to return.
    #[serde(default = "_default_max_results")]
    pub max_results: usize,
    /// Whether to include HPO terms.
    #[serde(default = "_default_hpo_terms")]
    pub hpo_terms: bool,
}

/// Return default of `Request::max_results`.
fn _default_max_results() -> usize {
    100
}

/// Return default of `Request::hpo_terms`.
fn _default_hpo_terms() -> bool {
    false
}

/// Result entry for `handle`.
#[derive(serde::Serialize, serde::Deserialize, utoipa::ToSchema, Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
#[serde_with::skip_serializing_none]
struct ResultEntry {
    /// The gene's NCBI ID.
    pub gene_ncbi_id: u32,
    /// The gene's HGNC symbol.
    pub gene_symbol: String,
    /// The gene's HGNC ID.
    pub hgnc_id: Option<String>,
    /// The gene's associated HPO terms.
    #[serde(default = "Option::default", skip_serializing_if = "Option::is_none")]
    pub hpo_terms: Option<Vec<ResultHpoTerm>>,
}

impl ResultEntry {
    pub fn from_gene_with_ontology(
        gene: &Gene,
        ontology: &Ontology,
        hpo_terms: bool,
        ncbi_to_hgnc: &HashMap<u32, String>,
    ) -> Self {
        let hpo_terms = if hpo_terms {
            let mut terms = gene
                .to_hpo_set(ontology)
                .child_nodes()
                .into_iter()
                .map(|term| ResultHpoTerm {
                    term_id: term.id().to_string(),
                    name: term.name().to_string(),
                })
                .collect::<Vec<_>>();
            terms.sort();
            Some(terms)
        } else {
            None
        };
        ResultEntry {
            gene_ncbi_id: gene.id().as_u32(),
            gene_symbol: gene.name().to_string(),
            hgnc_id: ncbi_to_hgnc.get(&gene.id().as_u32()).cloned(),
            hpo_terms,
        }
    }
}

/// Container for the result.
#[derive(Debug, serde::Serialize, serde::Deserialize, utoipa::ToSchema)]
struct Container {
    /// Version information.
    pub version: crate::common::Version,
    /// The original query records.
    pub query: Query,
    /// The resulting records for the scored genes.
    pub result: Vec<ResultEntry>,
}

/// Query for genes in the HPO database.
#[allow(clippy::unused_async)]
#[utoipa::path(
    responses(
        (status = 200, description = "The query was successful.", body = Container),
    )
)]
#[get("/hpo/genes")]
async fn handle(
    data: Data<WebServerData>,
    _path: Path<()>,
    query: web::Query<Query>,
) -> actix_web::Result<impl Responder, CustomError> {
    let ontology = &data.ontology;
    let match_ = query.match_.unwrap_or_default();
    let mut result: Vec<ResultEntry> = Vec::new();

    if match_ == Match::Exact {
        let gene = if let Some(gene_id) = &query.gene_id {
            let gene_id = if let Ok(ncbi_gene_id) = gene_id.parse::<u32>() {
                Ok(GeneId::from(ncbi_gene_id))
            } else if let Some(ncbi_gene_id) = data.hgnc_to_ncbi.get(gene_id) {
                Ok(GeneId::from(*ncbi_gene_id))
            } else {
                Err(CustomError::new(anyhow::anyhow!("could not parse gene ID")))
            }?;
            ontology.gene(&gene_id)
        } else if let Some(gene_symbol) = &query.gene_symbol {
            ontology.gene_by_name(gene_symbol)
        } else {
            None
        };
        if let Some(gene) = gene {
            result.push(ResultEntry::from_gene_with_ontology(
                gene,
                ontology,
                query.hpo_terms,
                &data.ncbi_to_hgnc,
            ));
        }
    } else if let Some(gene_symbol) = &query.gene_symbol {
        let mut it = ontology.genes();
        let mut gene = it.next();
        while gene.is_some() && result.len() < query.max_results {
            let symbol = gene.expect("checked above").symbol();
            let is_match = match query.match_.unwrap_or_default() {
                Match::Prefix => symbol.starts_with(gene_symbol),
                Match::Suffix => symbol.ends_with(gene_symbol),
                Match::Contains => symbol.contains(gene_symbol),
                Match::Exact => panic!("cannot happen here"),
            };
            if is_match {
                result.push(ResultEntry::from_gene_with_ontology(
                    gene.expect("checked above"),
                    ontology,
                    query.hpo_terms,
                    &data.ncbi_to_hgnc,
                ));
            }

            gene = it.next();
        }
    }

    result.sort();

    let result = Container {
        version: crate::common::Version::new(&data.ontology.hpo_version()),
        query: query.into_inner(),
        result,
    };

    Ok(Json(result))
}

#[cfg(test)]
mod test {
    /// Helper function for running a query.
    #[allow(dead_code)]
    async fn run_query(uri: &str) -> Result<super::Container, anyhow::Error> {
        let ontology = crate::common::load_hpo("tests/data/hpo")?;
        let ncbi_to_hgnc =
            crate::common::hgnc_xlink::load_ncbi_to_hgnc("tests/data/hgnc_xlink.tsv")?;
        let hgnc_to_ncbi = crate::common::hgnc_xlink::inverse_hashmap(&ncbi_to_hgnc);
        let hpo_doc = fastobo::from_file("tests/data/hpo/hp.obo")?;
        let app = actix_web::test::init_service(
            actix_web::App::new()
                .app_data(actix_web::web::Data::new(
                    crate::server::run::WebServerData {
                        ontology,
                        ncbi_to_hgnc,
                        hgnc_to_ncbi,
                        full_text_index: crate::index::Index::new(hpo_doc)?,
                    },
                ))
                .service(super::handle),
        )
        .await;
        let req = actix_web::test::TestRequest::get().uri(uri).to_request();
        let resp: super::Container = actix_web::test::call_and_read_body_json(&app, req).await;

        Ok(resp)
    }

    #[actix_web::test]
    async fn hpo_genes_ncbi_gene_id_exact_no_hpo_terms() -> Result<(), anyhow::Error> {
        Ok(insta::assert_yaml_snapshot!(
            &run_query("/hpo/genes?gene_id=2348").await?
        ))
    }

    #[actix_web::test]
    async fn hpo_genes_ncbi_gene_id_exact_with_hpo_terms() -> Result<(), anyhow::Error> {
        Ok(insta::assert_yaml_snapshot!(
            &run_query("/hpo/genes?gene_id=2348&hpo_terms=true").await?
        ))
    }

    #[actix_web::test]
    async fn hpo_genes_hgnc_gene_id_exact_no_hpo_terms() -> Result<(), anyhow::Error> {
        Ok(insta::assert_yaml_snapshot!(
            &run_query("/hpo/genes?gene_id=HGNC:3791").await?
        ))
    }

    #[actix_web::test]
    async fn hpo_genes_hgnc_gene_id_exact_with_hpo_terms() -> Result<(), anyhow::Error> {
        Ok(insta::assert_yaml_snapshot!(
            &run_query("/hpo/genes?gene_id=HGNC:3791&hpo_terms=true").await?
        ))
    }

    #[actix_web::test]
    async fn hpo_genes_gene_symbol_exact_no_hpo_terms() -> Result<(), anyhow::Error> {
        Ok(insta::assert_yaml_snapshot!(
            &run_query("/hpo/genes?gene_symbol=TGDS").await?
        ))
    }

    #[actix_web::test]
    async fn hpo_genes_gene_symbol_exact_with_hpo_terms() -> Result<(), anyhow::Error> {
        Ok(insta::assert_yaml_snapshot!(
            &run_query("/hpo/genes?gene_symbol=TGDS&hpo_terms=true").await?
        ))
    }

    #[actix_web::test]
    async fn hpo_genes_gene_symbol_prefix_no_hpo_terms() -> Result<(), anyhow::Error> {
        Ok(insta::assert_yaml_snapshot!(
            &run_query("/hpo/genes?gene_symbol=TGD&match=prefix").await?
        ))
    }

    #[actix_web::test]
    async fn hpo_genes_gene_symbol_prefix_with_hpo_terms() -> Result<(), anyhow::Error> {
        Ok(insta::assert_yaml_snapshot!(
            &run_query("/hpo/genes?gene_symbol=TGD&match=prefix&hpo_terms=true").await?
        ))
    }

    #[actix_web::test]
    async fn hpo_genes_gene_symbol_suffix_no_hpo_terms() -> Result<(), anyhow::Error> {
        Ok(insta::assert_yaml_snapshot!(
            &run_query("/hpo/genes?gene_symbol=GDS&match=suffix").await?
        ))
    }

    #[actix_web::test]
    async fn hpo_genes_gene_symbol_suffix_with_hpo_terms() -> Result<(), anyhow::Error> {
        Ok(insta::assert_yaml_snapshot!(
            &run_query("/hpo/genes?gene_symbol=GDS&match=suffix&hpo_terms=true").await?
        ))
    }

    #[actix_web::test]
    async fn hpo_genes_gene_symbol_contains_no_hpo_terms() -> Result<(), anyhow::Error> {
        Ok(insta::assert_yaml_snapshot!(
            &run_query("/hpo/genes?gene_symbol=GD&match=contains").await?
        ))
    }

    #[actix_web::test]
    async fn hpo_genes_gene_symbol_contains_with_hpo_terms() -> Result<(), anyhow::Error> {
        Ok(insta::assert_yaml_snapshot!(
            &run_query("/hpo/genes?gene_symbol=GD&match=contains&hpo_terms=true").await?
        ))
    }
}
