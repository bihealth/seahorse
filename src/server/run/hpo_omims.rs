//! Implementation of `/hpo/omims`.

use actix_web::{
    get,
    web::{self, Data, Json, Path},
    Responder,
};
use hpo::{
    annotations::{OmimDisease, OmimDiseaseId},
    term::HpoGroup,
    Ontology,
};

use crate::{common::Version, server::run::WebServerData};

use super::{CustomError, Match, ResultHpoTerm};

/// Parameters for `handle`.
///
/// This allows to query for diseases.  The first given of the following
/// is interpreted.
///
/// - `omim_id` -- specify disease ID
/// - `name` -- specify the name to query for
/// - `max_results` -- the maximum number of records to return
/// - `hpo_terms` -- whether to include `"hpo_terms"` in result
///
/// The following propery defines how matches are performed:
///
/// - `match` -- how to match
#[derive(
    serde::Serialize, serde::Deserialize, utoipa::ToSchema, utoipa::IntoParams, Debug, Clone,
)]
#[schema(title = "HpoOmimsQuery")]
pub struct Query {
    /// The OMIM ID to search for.
    pub omim_id: Option<String>,
    /// The disease name to search for.
    pub name: Option<String>,
    /// The match mode.
    #[serde(alias = "match")]
    pub match_: Option<Match>,
    /// Maximal number of results to return.
    #[serde(default = "_default_max_results")]
    pub max_results: usize,
    /// Whether to include HPO terms.
    #[serde(default = "_default_hpo_terms")]
    pub hpo_terms: bool,
}

/// Return default of `Request::max_results`.
fn _default_max_results() -> usize {
    100
}

/// Return default of `Request::hpo_terms`.
fn _default_hpo_terms() -> bool {
    false
}

/// Result entry for `handle`.
#[derive(serde::Serialize, serde::Deserialize, utoipa::ToSchema, Debug, Clone)]
#[schema(title = "HpoOmimsResultEntry")]
pub struct ResultEntry {
    /// The OMIM ID.
    pub omim_id: String,
    /// The OMIM disease name.
    pub name: String,
    /// The gene's associated HPO terms.
    #[serde(default = "Option::default", skip_serializing_if = "Option::is_none")]
    pub hpo_terms: Option<Vec<ResultHpoTerm>>,
}

impl PartialEq for ResultEntry {
    fn eq(&self, other: &Self) -> bool {
        (self.omim_id == other.omim_id) && (self.name == other.name)
    }
}

impl Eq for ResultEntry {}

impl PartialOrd for ResultEntry {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for ResultEntry {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        match self.omim_id.cmp(&other.omim_id) {
            core::cmp::Ordering::Equal => {}
            ord => return ord,
        }
        self.name.cmp(&other.name)
    }
}

impl ResultEntry {
    /// Create a `ResultEntry` from an `OmimDisease`.
    pub fn from_omim_disease_with_ontology(
        omim_disease: &OmimDisease,
        ontology: &Ontology,
        hpo_terms: bool,
    ) -> Self {
        let hpo_terms = if hpo_terms {
            let mut result = omim_disease
                .to_hpo_set(ontology)
                .child_nodes()
                .into_iter()
                .collect::<HpoGroup>()
                .into_iter()
                .filter_map(|term_id| ontology.hpo(term_id))
                .map(|term| ResultHpoTerm {
                    term_id: term.id().to_string(),
                    name: term.name().to_string(),
                })
                .collect::<Vec<_>>();
            result.sort();
            Some(result)
        } else {
            None
        };
        ResultEntry {
            omim_id: omim_disease.id().to_string(),
            name: omim_disease.name().to_string(),
            hpo_terms,
        }
    }
}

/// Container for the result.
#[derive(serde::Serialize, serde::Deserialize, utoipa::ToSchema, Debug)]
#[schema(title = "HpoOmimsResult")]
pub struct Result {
    /// Version information.
    pub version: Version,
    /// The original query records.
    pub query: Query,
    /// The resulting records for the scored genes.
    pub result: Vec<ResultEntry>,
}

/// Query for OMIM diseases in the HPO database.
#[allow(clippy::unused_async)]
#[utoipa::path(
    params(Query),
    responses(
        (status = 200, description = "The query was successful.", body = Result),
    )
)]
#[get("/hpo/omims")]
async fn handle(
    data: Data<WebServerData>,
    _path: Path<()>,
    query: web::Query<Query>,
) -> actix_web::Result<impl Responder, CustomError> {
    let ontology = &data.ontology;
    let match_ = query.match_.unwrap_or_default();
    let mut result: Vec<ResultEntry> = Vec::new();

    if match_ == Match::Exact {
        let omim_disease = if let Some(omim_id) = &query.omim_id {
            let omim_id = OmimDiseaseId::try_from(omim_id.as_ref())
                .map_err(|e| CustomError::new(anyhow::anyhow!(e)))?;
            ontology.omim_disease(&omim_id)
        } else if let Some(name) = &query.name {
            let mut omim_disease = None;
            let mut it = ontology.omim_diseases();
            let mut tmp = it.next();
            while tmp.is_some() && omim_disease.is_none() {
                if tmp.expect("checked above").name() == name {
                    omim_disease = tmp;
                }
                tmp = it.next();
            }
            omim_disease
        } else {
            None
        };
        if let Some(omim_disease) = &omim_disease {
            result.push(ResultEntry::from_omim_disease_with_ontology(
                omim_disease,
                ontology,
                query.hpo_terms,
            ));
        }
    } else if let Some(name) = &query.name {
        let mut it = ontology.omim_diseases();
        let mut omim_disease = it.next();
        while omim_disease.is_some() && result.len() < query.max_results {
            let omim_name = omim_disease.as_ref().expect("checked above").name();
            let is_match = match query.match_.unwrap_or_default() {
                Match::Prefix => omim_name.starts_with(name),
                Match::Suffix => omim_name.ends_with(name),
                Match::Contains => omim_name.contains(name),
                Match::Exact => panic!("cannot happen here"),
            };
            if is_match {
                result.push(ResultEntry::from_omim_disease_with_ontology(
                    omim_disease.as_ref().expect("checked above"),
                    ontology,
                    query.hpo_terms,
                ));
            }

            omim_disease = it.next();
        }
    }

    result.sort();

    let result = Result {
        version: Version::new(&data.ontology.hpo_version()),
        query: query.into_inner(),
        result,
    };

    Ok(Json(result))
}

#[cfg(test)]
mod test {
    /// Helper function for running a query.
    #[allow(dead_code)]
    async fn run_query(uri: &str) -> Result<super::Result, anyhow::Error> {
        let ontology = crate::common::load_hpo("tests/data/hpo")?;
        let ncbi_to_hgnc =
            crate::common::hgnc_xlink::load_ncbi_to_hgnc("tests/data/hgnc_xlink.tsv")?;
        let hgnc_to_ncbi = crate::common::hgnc_xlink::inverse_hashmap(&ncbi_to_hgnc);
        let hpo_doc = fastobo::from_file("tests/data/hpo/hp.obo")?;
        let app = actix_web::test::init_service(
            actix_web::App::new()
                .app_data(actix_web::web::Data::new(
                    crate::server::run::WebServerData {
                        ontology,
                        ncbi_to_hgnc,
                        hgnc_to_ncbi,
                        full_text_index: crate::index::Index::new(hpo_doc)?,
                    },
                ))
                .service(super::handle),
        )
        .await;
        let req = actix_web::test::TestRequest::get().uri(uri).to_request();
        let resp: super::Result = actix_web::test::call_and_read_body_json(&app, req).await;

        Ok(resp)
    }

    #[actix_web::test]
    async fn hpo_omims_omim_id_exact_no_hpo_terms() -> Result<(), anyhow::Error> {
        Ok(insta::assert_yaml_snapshot!(
            &run_query("/hpo/omims?omim_id=616145").await?
        ))
    }

    #[actix_web::test]
    async fn hpo_omims_omim_id_exact_with_hpo_terms() -> Result<(), anyhow::Error> {
        Ok(insta::assert_yaml_snapshot!(
            &run_query("/hpo/omims?omim_id=616145&hpo_terms=true").await?
        ))
    }

    #[actix_web::test]
    async fn hpo_omims_name_exact_no_hpo_terms() -> Result<(), anyhow::Error> {
        Ok(insta::assert_yaml_snapshot!(
            &run_query("/hpo/omims?name=Catel-Manzke+syndrome").await?
        ))
    }

    #[actix_web::test]
    async fn hpo_omims_name_exact_with_hpo_terms() -> Result<(), anyhow::Error> {
        Ok(insta::assert_yaml_snapshot!(
            &run_query("/hpo/omims?name=Catel-Manzke+syndrome&hpo_terms=true").await?
        ))
    }

    #[actix_web::test]
    async fn hpo_omims_name_prefix_no_hpo_terms() -> Result<(), anyhow::Error> {
        Ok(insta::assert_yaml_snapshot!(
            &run_query("/hpo/omims?name=Catel-Manzke+syndro&match=prefix").await?
        ))
    }

    #[actix_web::test]
    async fn hpo_omims_name_prefix_with_hpo_terms() -> Result<(), anyhow::Error> {
        Ok(insta::assert_yaml_snapshot!(
            &run_query("/hpo/omims?name=Catel-Manzke+syndro&match=prefix&hpo_terms=true").await?
        ))
    }

    #[actix_web::test]
    async fn hpo_omims_name_suffix_no_hpo_terms() -> Result<(), anyhow::Error> {
        Ok(insta::assert_yaml_snapshot!(
            &run_query("/hpo/omims?name=tel-Manzke+syndrome&match=suffix").await?
        ))
    }

    #[actix_web::test]
    async fn hpo_omims_name_suffix_with_hpo_terms() -> Result<(), anyhow::Error> {
        Ok(insta::assert_yaml_snapshot!(
            &run_query("/hpo/omims?name=tel-Manzke+syndrome&match=suffix&hpo_terms=true").await?
        ))
    }

    #[actix_web::test]
    async fn hpo_omims_name_contains_no_hpo_terms() -> Result<(), anyhow::Error> {
        Ok(insta::assert_yaml_snapshot!(
            &run_query("/hpo/omims?name=tel-Manzke+syndro&match=contains").await?
        ))
    }

    #[actix_web::test]
    async fn hpo_omims_name_contains_with_hpo_terms() -> Result<(), anyhow::Error> {
        Ok(insta::assert_yaml_snapshot!(
            &run_query("/hpo/omims?name=tel-Manzke+syndro&match=contains&hpo_terms=true").await?
        ))
    }
}
