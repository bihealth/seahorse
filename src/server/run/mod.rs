//! Implementation of the Actix server.

pub mod hpo_genes;
pub mod hpo_omims;
pub mod hpo_sim;
pub mod hpo_terms;

use std::collections::HashMap;

use actix_web::{middleware::Logger, web::Data, App, HttpServer, ResponseError};
use serde::{Deserialize, Deserializer, Serialize};

use crate::common::load_hpo;

/// Data structure for the web server data.
pub struct WebServerData {
    /// The HPO ontology (`hpo` crate).
    pub ontology: hpo::Ontology,
    /// Xlink map from NCBI gene ID to HGNC gene ID.
    pub ncbi_to_hgnc: HashMap<u32, String>,
    /// Xlink map from HGNC gene ID to NCBI gene ID.
    pub hgnc_to_ncbi: HashMap<String, u32>,
    /// The full text index over the HPO OBO document.
    pub full_text_index: crate::index::Index,
}

/// Command line arguments for `server pheno` sub command.
#[derive(clap::Parser, Debug)]
#[command(author, version, about = "Run viguno REST API server", long_about = None)]
pub struct Args {
    /// Path to the directory with the HPO files.
    #[arg(long, required = true)]
    pub path_hpo_dir: String,
    /// Path to the TSV file with the HGNC xlink data.
    #[arg(long, required = true)]
    pub path_hgnc_xlink: String,

    /// Whether to suppress printing hints.
    #[arg(long, default_value_t = false)]
    pub suppress_hints: bool,

    /// IP to listen on.
    #[arg(long, default_value = "127.0.0.1")]
    pub listen_host: String,
    /// Port to listen on.
    #[arg(long, default_value_t = 8080)]
    pub listen_port: u16,
}

#[derive(Debug)]
struct CustomError {
    err: anyhow::Error,
}

impl std::fmt::Display for CustomError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:?}", self.err)
    }
}

impl CustomError {
    fn new(err: anyhow::Error) -> Self {
        CustomError { err }
    }
}

impl ResponseError for CustomError {}

/// Specify how to perform query matches in the API calls.
#[derive(Serialize, Deserialize, Debug, Clone, Copy, Default, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
enum Match {
    #[default]
    Exact,
    Prefix,
    Suffix,
    Contains,
}

/// Representation of a gene.
#[derive(
    serde::Deserialize, serde::Serialize, Default, Debug, Clone, PartialOrd, Ord, PartialEq, Eq,
)]
#[serde_with::skip_serializing_none]
struct ResultGene {
    /// The HPO ID.
    pub ncbi_gene_id: u32,
    /// The description.
    pub gene_symbol: String,
    /// The HGNC ID.
    pub hgnc_id: Option<String>,
}

/// Representation of an HPO term.
#[derive(
    serde::Deserialize, serde::Serialize, Default, Debug, Clone, PartialOrd, Ord, PartialEq, Eq,
)]
struct ResultHpoTerm {
    /// The HPO ID.
    pub term_id: String,
    /// The term name.
    pub name: String,
}

/// Helper to deserialize a comma-separated list of strings.
fn vec_str_deserialize<'de, D>(deserializer: D) -> Result<Vec<String>, D::Error>
where
    D: Deserializer<'de>,
{
    let str_sequence = String::deserialize(deserializer)?;
    Ok(str_sequence
        .split(',')
        .map(std::borrow::ToOwned::to_owned)
        .collect())
}

/// Helper to deserialize a comma-separated list of strings.
fn option_vec_str_deserialize<'de, D>(deserializer: D) -> Result<Option<Vec<String>>, D::Error>
where
    D: Deserializer<'de>,
{
    let str_sequence = String::deserialize(deserializer)?;
    if str_sequence.is_empty() {
        Ok(None)
    } else {
        Ok(Some(
            str_sequence
                .split(',')
                .map(std::borrow::ToOwned::to_owned)
                .collect(),
        ))
    }
}

/// Main entry point for running the REST server.
#[allow(clippy::unused_async)]
#[actix_web::main]
pub async fn main(args: &Args, dbs: Data<WebServerData>) -> std::io::Result<()> {
    HttpServer::new(move || {
        App::new()
            .app_data(dbs.clone())
            .service(hpo_genes::handle)
            .service(hpo_terms::handle)
            .service(hpo_omims::handle)
            .service(hpo_sim::term_term::handle)
            .service(hpo_sim::term_gene::handle)
            .wrap(Logger::default())
    })
    .bind((args.listen_host.as_str(), args.listen_port))?
    .run()
    .await
}

/// Print some hints via `tracing::info!`.
pub fn print_hints(args: &Args) {
    tracing::info!(
        "Launching server main on http://{}:{} ...",
        args.listen_host.as_str(),
        args.listen_port
    );

    // Short-circuit if no hints are to be
    if args.suppress_hints {
        return;
    }

    // The endpoint `/hpo/genes` provides information related to genes by symbol.
    tracing::info!(
        "  try: http://{}:{}/hpo/genes?gene_symbol=TGDS",
        args.listen_host.as_str(),
        args.listen_port
    );
    // Also, you can query `/hpo/genes` by NCBI gene ID and return the HPO terms of the gene.
    tracing::info!(
        "  try: http://{}:{}/hpo/genes?gene_id=23483&hpo_terms=true",
        args.listen_host.as_str(),
        args.listen_port
    );
    // The `/hpo/omims` term provides information on OMIM terms and can include HPO terms for
    // the disease.
    tracing::info!(
        "  try: http://{}:{}/hpo/omims?omim_id=616145&hpo_terms=true",
        args.listen_host.as_str(),
        args.listen_port
    );
    // The `/hpo/terms` endpoint allows to query by HPO term ID and optionally return a list of
    // genes that are linked to the term.
    tracing::info!(
        "  try: http://{}:{}/hpo/terms?term_id=HP:0000023&genes=true",
        args.listen_host.as_str(),
        args.listen_port
    );
    // We can use `/hpo/sim/term-term` to compute similarity between two HPO term sets `lhs`
    // and `rhs` using a similarity metric.
    tracing::info!(
        "  try: http://{}:{}/hpo/sim/term-term?lhs=HP:0001166,HP:0040069&rhs=HP:0005918,\
        HP:0004188",
        args.listen_host.as_str(),
        args.listen_port
    );
    // The endpoint `/hpo/sim/term-gene` allows to compute the same for a list of `terms` and
    // `gene_symbols`.
    tracing::info!(
        "  try: http://{}:{}/hpo/sim/term-gene?terms=HP:0001166,HP:0000098&gene_symbols=FBN1,TGDS,TTN",
        args.listen_host.as_str(),
        args.listen_port
    );
}

/// Main entry point for `run-server` sub command.
///
/// # Errors
///
/// In the case that there is an error running the server.
pub fn run(args_common: &crate::common::Args, args: &Args) -> Result<(), anyhow::Error> {
    tracing::info!("args_common = {:?}", &args_common);
    tracing::info!("args = {:?}", &args);

    if let Some(level) = args_common.verbose.log_level() {
        match level {
            log::Level::Trace | log::Level::Debug => {
                std::env::set_var("RUST_LOG", "debug");
                env_logger::init_from_env(env_logger::Env::new().default_filter_or("info"));
            }
            _ => (),
        }
    }

    // Load data that we need for running the server.
    tracing::info!("Loading HPO...");
    let before_loading = std::time::Instant::now();
    let ontology = load_hpo(&args.path_hpo_dir)?;
    tracing::info!("...done loading HPO in {:?}", before_loading.elapsed());

    tracing::info!("Loading HGNC xlink...");
    let before_load_xlink = std::time::Instant::now();
    let ncbi_to_hgnc = crate::common::hgnc_xlink::load_ncbi_to_hgnc(&args.path_hgnc_xlink)?;
    let hgnc_to_ncbi = crate::common::hgnc_xlink::inverse_hashmap(&ncbi_to_hgnc);
    tracing::info!(
        "... done loading HGNC xlink in {:?}",
        before_load_xlink.elapsed()
    );

    tracing::info!("Loading HPO OBO...");
    let before_load_obo = std::time::Instant::now();
    let hpo_doc = fastobo::from_file(format!("{}/{}", &args.path_hpo_dir, "hp.obo"))
        .map_err(|e| anyhow::anyhow!("Error loading HPO OBO: {}", e))?;
    tracing::info!(
        "... done loading HPO OBO in {:?}",
        before_load_obo.elapsed()
    );

    tracing::info!("Indexing OBO...");
    let before_index_obo = std::time::Instant::now();
    let full_text_index = crate::index::Index::new(hpo_doc)
        .map_err(|e| anyhow::anyhow!("Error indexing HPO OBO: {}", e))?;
    tracing::info!("... done indexing OBO in {:?}", before_index_obo.elapsed());

    let data = actix_web::web::Data::new(WebServerData {
        ontology,
        ncbi_to_hgnc,
        hgnc_to_ncbi,
        full_text_index,
    });

    // Print the server URL and some hints (the latter: unless suppressed).
    print_hints(args);
    // Launch the Actix web server.
    main(args, data)?;

    tracing::info!("All done. Have a nice day!");
    Ok(())
}
