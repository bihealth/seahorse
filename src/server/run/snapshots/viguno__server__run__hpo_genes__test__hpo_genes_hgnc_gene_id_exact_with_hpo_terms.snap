---
source: src/server/run/hpo_genes.rs
expression: "&run_query(\"/hpo/genes?gene_id=HGNC:3791&hpo_terms=true\").await?"
---
version:
  hpo: 2023-04-05
  viguno: 0.0.0
query:
  gene_id: "HGNC:3791"
  gene_symbol: ~
  match_: ~
  max_results: 100
  hpo_terms: true
result:
  - gene_ncbi_id: 2348
    gene_symbol: FOLR1
    hgnc_id: "HGNC:3791"
    hpo_terms:
      - term_id: "HP:0000007"
        name: Autosomal recessive inheritance
      - term_id: "HP:0001249"
        name: Intellectual disability
      - term_id: "HP:0001250"
        name: Seizure
      - term_id: "HP:0002180"
        name: Neurodegeneration
      - term_id: "HP:0002376"
        name: Developmental regression
